### ISSUE TITLE
Feature Request: built-in markdown editor for Readme.md in Edit mode

### ISSUE THREAD

**Issue by zhiyuli** [hydroshare/hydroshare#3731]
I request for integration of a built-in markdown editor for the readme.md in Edit mode. 

Currently I have to edit a readme.md somewhere else, then upload/re-upload to HydroShare resource to see if it displays as expected, which usually iterates several times.

It would simplify this process a lot if there is a build-in markdown editor for the readme.md file in Edit mode on the landing page.

Also with a build-in editor, user can clearly know what markdown syntax/directives are supported or not. See this related feature request: #3728

**Comment by Maurier** [comment]
I think this could be implemented into the site the same way GitHub does it (two tabs, text and preview). Further more, this is probably easy to implement since we can add an endpoint that takes in the text, passes it to the markdown pluggin in the backend, and returns the markdown to render to the frontend.

**Comment by horsburgh** [comment]
@Maurier - I like this idea, and was hoping we would get to it eventually after we did the initial implementation for view of readme files. We will also want to add something to the interface (button?) to make sure users can create a readme without uploading a file first.

**Comment by Maurier** [comment]
@horsburgh - Correct. Once these ideas get implemented, a button to create the readme directly in the browser seems like a logical addition.

**Comment by hyi** [comment]
@Maurier @horsburgh I like this idea also and it looks like it is not hard to implement at least for @Maurier. I think this will be a useful feature appreciated by users. On this note, I am thinking whether we could expand this a bit by not only allowing to display and edit readme.md but also allowing to display and edit other markdown files in a resource in a similar way, e.g., when a user selects another markdown file included in a resource. One use case for this could be illustrated in this evolving publication list resource https://www.hydroshare.org/resource/ceaffb1773dd4653a23461ec2d4ba01e/ I created to potentially address a request from @ChristinaB. The intend is for hydroshare team members to collectively edit this list to make it evolving and up to date. Since we currently only allow to display readme.md, I put the publication list in readme.md for now so that it can be displayed on resource landing page just for illustration, but if we could support displaying/editing any markdown files in a resource, this publication list should go to a separate markdown file, e.g., hs_publication_list.md. Hope this makes sense.

**Comment by horsburgh** [comment]
@hyi - I can see the use case there. That would be a larger change that would take a little additional thinking because it is possible to have multiple ".md" or ".txt" files in one folder in a resource. GitHub handles this by only rendering the "readme.md" file on their equivalent of the landing page, but they provide a view/edit page for individual, editable files (e.g., .md, .txt, code files, etc.). 

I like the functionality of having a single defined readme.md file in the root folder of the resource and predictable and set behavior for display of that on the landing page. We might extend this to the ability to have a readme.md file in any folder in the resource that gets rendered on the resource landing page as a user navigates from folder to folder (as GitHub does). 

If we were to do a general markdown editor, I think HydroShare should provide the user with a separate page for editing a file (as GitHub does) so that it can be handled generically.

**Comment by hyi** [comment]
@horsburgh Yeah, I can see the complexity involved mainly on UI of resource landing page in order to support this. Agreed that displaying and editing readme.md by default on resource landing page is the way to go. Supporting a general markdown editor for markdown files included in a resource could be in a separate page as you suggested, or perhaps adding an edit right click menu item when a user right click on a markdown file that brings up a modal dialog for displaying/editing or something like that.
